File: 03. Stack and Queue/03. Next Greater Element Circular/main.py
def next_greater_element_circular(nums):
    n = len(nums)
    result = [-1] * n
    stack = []

    for i in range(2 * n):
        while stack and nums[stack[-1]] < nums[i % n]:
            result[stack.pop()] = nums[i % n]
        if i < n:
            stack.append(i)

    return result

# Main function
if __name__ == "__main__":
    nums = [4, 5, 2, 10, 8]
    result = next_greater_element_circular(nums)

    print("Next Greater Elements (Circular):")
    print(result)

File: 03. Stack and Queue/03. Next Greater Element Circular/main.java
import java.util.*;

public class main {
    public static int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < 2 * n; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i % n]) {
                result[stack.pop()] = nums[i % n];
            }
            if (i < n) stack.push(i);
        }

        return result;
    }

    public static void main(String[] args) {
        int[] nums = {4, 5, 2, 10, 8};
        int[] result = nextGreaterElements(nums);

        System.out.println("Next Greater Elements (Circular):");
        System.out.println(Arrays.toString(result));
    }
}
